src/runtime/relax_vm/nd_allocator.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/runtime/relax_vm/nd_allocator.h
+ */
+#ifndef TVM_RUNTIME_RELAX_VM_ND_ALLOCATOR_H_
+#define TVM_RUNTIME_RELAX_VM_ND_ALLOCATOR_H_
+
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/relax_vm/memory_manager.h>
+
+#include <atomic>
+
+namespace tvm {
+namespace runtime {
+namespace relax_vm {
+
+class NDAllocator final : public Allocator {
+ public:
+  explicit NDAllocator(Device dev) : Allocator(kNd), used_memory_(0), device_(dev) {}
+
+  Buffer Alloc(ShapeTuple shape, DLDataType type_hint, String mem_scope) override {
+    DLTensor temp;
+    temp.data = nullptr;
+    temp.device = device_;
+    temp.ndim = shape.size();
+    temp.dtype = type_hint;
+    temp.shape = const_cast<int64_t*>(shape.data());
+    temp.strides = nullptr;
+    temp.byte_offset = 0;
+    size_t nbytes = GetDataSize(temp);
+
+    Buffer buf;
+    buf.device = device_;
+    buf.size = nbytes;
+    buf.data = runtime::DeviceAPI::Get(device_)->AllocDataSpace(device_, shape.size(), shape.data(),
+                                                                type_hint, mem_scope);
+    used_memory_.fetch_add(nbytes, std::memory_order_relaxed);
+    DLOG(INFO) << "allocate " << nbytes << " B, used memory " << used_memory_ << " B";
+    return buf;
+  }
+
+  Buffer Alloc(size_t nbytes, size_t alignment, DLDataType type_hint) override {
+    Buffer buf;
+    buf.device = device_;
```

Comment:
seems we don't need ND Allocator, just can let Naive allocator redirects to device API. We might need some better strategy later for things like hexagon allocator incase caching or pooling logic is nedded

Reply:
Sorry, i was meant to comment on the nd version's `Alloc`. Seems that we can have both 1D and ND directly implemented in the NaiveAllocator atm.

Reply:
ah okay, yeah that makes sense. I'll get rid of the NDAllocator and implement ND version with NaiveAllocator.

Reply:
One problem I see though is that the difference in how `alloc_storage` with scope (which takes physical shape) differs from existing `alloc_storage` (that takes raw bytes) could cause confusion. Especially when we get an alloc_storage call with 1d shape, and global memory scope, do we consider the 1d shape as raw bytes, or just physical shape (number of elements)?

I think unifying this discrepancy one way or another is needed if we have to get rid of NDAllocator.

Reply:
I've removed `NDAllocator`. Thanks for the feedback.